### PR TITLE
Readonly fields have an accessible home when accessed from init-only setter

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/45657

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 1]
```diff
@@ -3360,19 +3360,18 @@ internal enum AddressKind
             }
 
             // while readonly fields have home it is not valid to refer to it when not constructing.
-            if (!TypeSymbol.Equals(field.ContainingType, method.ContainingType, TypeCompareKind.ConsiderEverything2))
+            if (!TypeSymbol.Equals(field.ContainingType, method.ContainingType, TypeCompareKind.ConsiderEverything))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs:3363]: >ConsiderEverything [](start = 96, length = 18)

This option doesn't look right, why would difference in nullable annotations, or dynamic, or native int, or tuple names should make a difference? I think we should ignore all options that could be ignored.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs:3363]: I think the problem can be demonstrated with this failing test:
```
        [Fact]
        public void Test()
        {
            string source = @"
public class C
{
    public static void Main()
    {
        System.Console.WriteLine(C1<int>.F1.content);
        System.Console.WriteLine(C2<int>.F1.content);
    }
}

public struct Container
{
    public int content;
}

class C1<T>
{
    public static readonly Container F1;
    
    static C1()
    {
        C1<T>.F1.content = 2;
    }
}


#nullable enable

class C2<T>
{
    public static readonly Container F1;
    
    static C2()
    {
        C2<T>.F1.content = 2;
    }
}
";
            var comp = CreateCompilation(new[] { source, IsExternalInitTypeDefinition }, parseOptions: TestOptions.RegularPreview, options: TestOptions.DebugExe);
            comp.VerifyEmitDiagnostics();
            CompileAndVerify(comp, expectedOutput: 
@"
2
2
");
        }
```

---
In reply to: [460448404](https://github.com/dotnet/roslyn/pull/46323#discussion_r460448404) [](ancestors = 460448404)

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: Thanks much for the test scenario. Added and fixed.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs [hunk 2]
```diff
@@ -2373,6 +2373,52 @@ public int Property
...
+
+    static C2()
+    {
+        C2<T>.F1.content = 3;
+    }
+}
+";
+            var comp = CreateCompilation(new[] { source, IsExternalInitTypeDefinition }, parseOptions: TestOptions.RegularPreview, options: TestOptions.DebugExe);
+            comp.VerifyEmitDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "2 3", verify: Verification.Skipped /* init-only */);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs:2419]: >verify: Verification.Skipped /* init-only */ [](start = 58, length = 44)

This isn't an init-only scenario, no need to skip verification, I think.

Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs]: I think you're right. I did run into a PE verification error though.

As far as I can tell we're emitting the same IL [already](https://sharplab.io/#v2:C4LglgNgPgAgzAAhgJgQYQLACgDe2EFLLL6F5aGVICMAdGgPYB2AzgxAKa0DqATmMA4AKNNQA8YJsAB8tAGJ0Axs0FSAlAG5SVGvWZtOPfoKEAiBKc3aqMOo1bsufAcLTIJU2QtrKpHdVoUhAC+2KFY2PAILMC8AK6KwOgqAIaSHLzY5JRRkkm+qsCB4ZGoomIAKtJZ1lG2AGwIvBwpACbMEACeyVJpTBkICoG11I2iQmrW2Tro4lXySir+SQC8CMiBlOElWADETHEQECkARpwI/qec2KXo7lU1QQR1o00t7UxdPcB9A0M3TxoY2QEym1kobkqXkWfikCDWcE2ITC2CAA===) and in non-generic scenario [too](https://sharplab.io/#v2:C4LglgNgPgAgzAAhgJgQYQLACgDe2EFKIwCMAbAgE4CmAhgCYD2AdhAJ7ovC1jPWUIAYiQDc+QqQpoAFAEpxBPFkIr0AOmFqAxl2rNgCALwJkY5YQC+ChNZTJrS1RJJq0LAM6MI1NQHVKYMDU0mgaLjr6esCyZipWWPHY8AjuwJQArloGbvo8fJTYjhKIvAYRQfpm8UA) (which also fails PE verification). 

Feels like a PE verification bug. Is there a set of IL verification rules that I could reference to confirm this?

```
            // [ : C::Main][mdToken=0x6000004][offset 0x00000001] Cannot change initonly field outside its .ctor.
            v.VerifyIL("C.Main", @"
{
  // Code size       45 (0x2d)
  .maxstack  1
  IL_0000:  nop
  IL_0001:  ldsflda    ""Container C1<int>.F1""
  IL_0006:  ldfld      ""int Container.content""
  IL_000b:  call       ""void System.Console.Write(int)""
  IL_0010:  nop
  IL_0011:  ldstr      "" ""
  IL_0016:  call       ""void System.Console.Write(string)""
  IL_001b:  nop
  IL_001c:  ldsflda    ""Container C2<int>.F1""
  IL_0021:  ldfld      ""int Container.content""
  IL_0026:  call       ""void System.Console.Write(int)""
  IL_002b:  nop
  IL_002c:  ret
}
");
 ```

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for second review. Thanks

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for second review. Thanks

File: src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs [hunk 3]
```diff
@@ -2437,8 +2505,8 @@ void M(C c)
 ";
 
             var comp = CreateCompilation(new[] { source, IsExternalInitTypeDefinition }, parseOptions: TestOptions.RegularPreview);
-            comp.VerifyDiagnostics(
-                // (9,9): error CS0191: A readonly field cannot be assigned to (except in a constructor or init-only setter of the type in which the field is defined or a variable initializer)
+            comp.VerifyEmitDiagnostics(
+                // (9,9): error CS0191: A readonly field cannot be assigned to (except in a constructor or init-only setter of the class in which the field is defined or a variable initializer))
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs:2509]: >class in which the field is defined or a variable initializer)) [](start = 131, length = 63)

Is this comment stale?

Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs:2509]: This is an artifact of the GitHub diff view, I think.

Locally, the file shows the correct comment:
```
                // (9,9): error CS0191: A readonly field cannot be assigned to (except in a constructor or init-only setter of the type in which the field is defined or a variable initializer)
                //         c.field = null; // 1
                Diagnostic(ErrorCode.ERR_AssgReadonly, "c.field").WithLocation(9, 9),
```

**Pull request merged**